pkg/rbac/rbac.go

```diff
@@ -0,0 +1,287 @@
+		Group:    "cluster.open-cluster-management.io",
+		Resource: "managedclusters",
+	},
+	verb: "metrics/",
+}
+
+// APIEngine is the rbac api for custom/finer grained RBAC, it should preferably be instantiated through
+// the NewAPIEngine function as it will do any required validation
+// Either or both Config and Client to connect to the Kubernetes cluster need to be configured.
+// any additional configuration needed later on e.g. settings for a cache etc would go in here as well
+
+// Note: Idea is that consumers of this API ( like Observability) will create a single instance of API Engine
+// with the KubeConfig of the Hub Cluster configured and whenever making a rbac api call for a specific user,
+// the  bearerToken of the user is input to the rbac api call. This bearerToken along with Hub cluster config
+// will be used to generate a kubeclient to connect to the cluster and get the rbac info for that user.
+//
+// However, KubeClient (instead of KubeConfig) may also be configured on the Rbac APIEngine,
+// this to allow for the cases where only a  kubeClient is available
+// or when rbac api calls are going to be made only for a single user or serviceaccount throughout.
+
+type APIEngine struct {
+	kubeConfig *rest.Config
+	kubeClient kubernetes.Interface
+}
+
+// kubeConfig  - Kubernetes Cluster configuration
+// kubeClient  - Client to talk to a Kubernetes cluster
+// Either or both Config and Client to connect to the Kubernetes cluster can be set.
+// throws an error when neither is set.
+func NewAPIEngine(kConfig *rest.Config, kClient kubernetes.Interface) (*APIEngine, error) {
+	if kClient == nil && kConfig == nil {
+		return nil, errors.New("either kubeConfig or kubeClient must be a non-nil value")
+	}
+
+	return &APIEngine{
+		kubeClient: kClient,
+		kubeConfig: kConfig,
+	}, nil
+}
+
+// returns the kubeclient to use to connect to the cluster when making rbac api calls
+// given a  valid userToken (bearerToken) and provided kubeConfig is configured  on the RbacEngine,
+// it generates a new kubeClient for that user
+// if not, any configured kubernetes client on the RbacEgine is returned
+func (r *APIEngine) getKubeClientForUser(userToken string) (kubernetes.Interface, error) {
+	if len(userToken) != 0 && r.kubeConfig != nil {
+		// ?? maybe I should make a copy of the kubeConfig everytime instead of setting it on the same copy
+		// ?? wondering if  tokens will overwrite each other when multiple calls  come in parallel
+		// ?? need some in parallel testing on live cluster
+		r.kubeConfig.BearerToken = userToken
+		// create the clientset
+		kclient, err := kubernetes.NewForConfig(r.kubeConfig)
+		if err != nil {
+			return nil, err
+		}
+
+		return kclient, nil
+	} else if r.kubeConfig == nil {
+		return r.kubeClient, nil
+	} else {
+		return nil, fmt.Errorf("failed to evaluate the acls for User")
+	}
+}
+
+// this API call returns the namespaces allowed for  metrics access for the  given list of managedcluster (or all)
```

Comment:
This is the common pattern in Go documentation.
```suggestion
// GetMetricsAccess returns the namespaces allowed for metrics access for the  given list of managedcluster (or all)
```